src/app/lib/checkbox/checkbox.component.ts

```diff
@@ -1,21 +1,158 @@
+  providers: [hcCheckboxValueAccessor],
+  exportAs: 'hcCheckbox'
 })
-export class CheckboxComponent implements OnInit {
-    @Input() name: string;
-    @Input() disabled: boolean = false;
-    @Input() model: boolean;
+export class CheckboxComponent implements ControlValueAccessor {
+  private uniqueId = `hc-checkbox-${nextCheckboxId++}`;
+  private _checked: boolean = false;
+  private _required: boolean = false;
+  private _disabled: boolean = false;
+  private _tabIndex: number;
+
+  // list out all properties or attributes an input[type=checkbox] can have since we're wrapping the input
+  @Input() value: string;
+  @Input() indeterminate: boolean;
+  @Input() id: string = this.uniqueId;
+  @Input() name: string | null = null;
+
+  @Output() change = new EventEmitter<CheckboxChangeEvent>();
+
+  @HostListener('blur')
+  onBlur() {
+    this.onTouchFunc();
+  }
+
+  @HostBinding('attr.id')
+  get getHostId(): string {
+    return this.uniqueId;
+  }
+
+  @Input()
+  get required(): boolean {
+    return this._required;
+  }
+
+  set required(required) {
+    this._required = this.anyToBoolean(required);
+  }
+
+  @Input()
+  get disabled(): boolean {
+    return this._disabled;
+  }
+
+  set disabled(isDisabled) {
+    this._disabled = this.anyToBoolean(isDisabled);
+  }
+
+  @Input()
+  get checked(): boolean {
+    return this._checked;
+  }
+
+  set checked(checked: boolean) {
+    if (this._checked === checked) {
+      return;
+    }
+    this._checked = checked;
+  }
 
-    constructor() {
+  get tabIndex(): number {
+    return this.disabled ? -1 : this._tabIndex;
+  }
+
+  set tabIndex(value: number) {
+    this._tabIndex = value == null ? 0 : value;
+  }
+
+  get inputId() {
+    if (this.id) {
+      return this.id;
     }
+    return `${this.uniqueId}-input`;
+  }
+
+  private onChangeFunc: (value: any) => void = () => {
+  };
+
+  private onTouchFunc: () => any = () => {
+  };
+
+  constructor(@Attribute('tabindex') tabindex: string) {
+    this.tabIndex = parseInt(tabindex, 10) || 0;
+  }
+
+  writeValue(value: any): void {
+    this.checked = !!value;
+  }
+
+  registerOnChange(fn: (value: any) => void): void {
+    this.onChangeFunc = fn;
+  }
+
+  registerOnTouched(fn: () => any): void {
+    this.onTouchFunc = fn;
+  }
 
-    ngOnInit() {
-        this.disabled = <any>this.disabled === 'true';
+  toggle() {
+    this.checked = !this.checked;
+  }
+
+  clickEvent(event: Event) {
+    event.stopPropagation(); // prevent native click event from being dispatched
+
+    if (!this.disabled) {
+      this.toggle();
+      this.emitChangeEvent();
+    }
+  }
+
+  stopChangeEvent(event: Event) {
+    event.stopPropagation(); // prevent native change event from emitting its own object through output 'change'
+  }
+
+  private emitChangeEvent(): void {
+    this.onChangeFunc(this.checked);
+    this.change.emit(new CheckboxChangeEvent(this, this.checked));
+  }
+
+  private anyToBoolean(value: any): boolean {
```

Comment:
Much better than my `<any>this.disabled === 'true'`

Reply:
I could see this being useful outside of this class as well; would it be good to extract this out as some sort of utility method?

Reply:
Exactly what I was thinking. I was going to do that in a different commit. I don't know what the best name would be for a folder and how we want to structure these util methods. I could just go with the generic folder name of util or core.